### ISSUE TITLE
add args to control padding in the definition of pretrained model in slim

### ISSUE THREAD

**Issue by zakizhou** [tensorflow/tensorflow#12268]
in some situation, the outputs of "same" and "valid" padding are same for original model, but when do fine-tuning with a new model, they are not,  adding an arg to control padding is meaningful.

**Comment by reedwm** [comment]
Can you clarify what you mean? What do you mean by adding an argument to control padding?

**Comment by zakizhou** [comment]
Thanks! @reedwm 

This is the source code of definition of vgg16 in slim:
```
def vgg_16(inputs,
           num_classes=1000,
           is_training=True,
           dropout_keep_prob=0.5,
           spatial_squeeze=True,
           scope='vgg_16'):
  """Oxford Net VGG 16-Layers version D Example.
  Note: All the fully_connected layers have been transformed to conv2d layers.
        To use in classification mode, resize input to 224x224.
  Args:
    inputs: a tensor of size [batch_size, height, width, channels].
    num_classes: number of predicted classes.
    is_training: whether or not the model is being trained.
    dropout_keep_prob: the probability that activations are kept in the dropout
      layers during training.
    spatial_squeeze: whether or not should squeeze the spatial dimensions of the
      outputs. Useful to remove unnecessary dimensions for classification.
    scope: Optional scope for the variables.
  Returns:
    the last op containing the log predictions and end_points dict.
  """
  with variable_scope.variable_scope(scope, 'vgg_16', [inputs]) as sc:
    end_points_collection = sc.original_name_scope + '_end_points'
    # Collect outputs for conv2d, fully_connected and max_pool2d.
    with arg_scope(
        [layers.conv2d, layers_lib.fully_connected, layers_lib.max_pool2d],
        outputs_collections=end_points_collection):
      net = layers_lib.repeat(
          inputs, 2, layers.conv2d, 64, [3, 3], scope='conv1')
      net = layers_lib.max_pool2d(net, [2, 2], scope='pool1')
      net = layers_lib.repeat(net, 2, layers.conv2d, 128, [3, 3], scope='conv2')
      net = layers_lib.max_pool2d(net, [2, 2], scope='pool2')
      net = layers_lib.repeat(net, 3, layers.conv2d, 256, [3, 3], scope='conv3')
      net = layers_lib.max_pool2d(net, [2, 2], scope='pool3')
      net = layers_lib.repeat(net, 3, layers.conv2d, 512, [3, 3], scope='conv4')
      net = layers_lib.max_pool2d(net, [2, 2], scope='pool4')
      net = layers_lib.repeat(net, 3, layers.conv2d, 512, [3, 3], scope='conv5')
      net = layers_lib.max_pool2d(net, [2, 2], scope='pool5')
      # Use conv2d instead of fully_connected layers.
      net = layers.conv2d(net, 4096, [7, 7], padding='VALID', scope='fc6')
      net = layers_lib.dropout(
          net, dropout_keep_prob, is_training=is_training, scope='dropout6')
      net = layers.conv2d(net, 4096, [1, 1], scope='fc7')
      net = layers_lib.dropout(
          net, dropout_keep_prob, is_training=is_training, scope='dropout7')
      net = layers.conv2d(
          net,
          num_classes, [1, 1],
          activation_fn=None,
          normalizer_fn=None,
          scope='fc8')
      # Convert end_points_collection into a end_point dict.
      end_points = utils.convert_collection_to_dict(end_points_collection)
      if spatial_squeeze:
        net = array_ops.squeeze(net, [1, 2], name='fc8/squeezed')
        end_points[sc.name + '/fc8'] = net
      return net, end_points
```
As you can see,  both functions(conv2d and maxpooling) use default padding argument which is "same" for conv2d and "valid" for maxpooling, for an input image with shape `[None, 224, 224, 3]` where 224 is a multiple of 32, "valid" padding of maxpooling has the same effect with "same" padding, but if the shape is `[None, 300, 300, 3]`(which is the input shape of ssd), output shape of "valid" padding is different from "same", so I want an argument in the definition of `vgg_16` to determine whether the padding for maxpooling is "same" or "valid".

**Comment by reedwm** [comment]
The `vgg_16` model assumes its input is 224x224. If you want to run it on images of a different size, you must first resize it to 224x224. Note the comment in the docstring which states "To use in classification mode, resize input to 224x224."

**Comment by zakizhou** [comment]
![image](http://joshua881228.webfactional.com/media/uploads/ReadingNote/arXiv_SSD/SSD.png)

ssd chooses vgg 16 as base network, if the padding is "valid" which is default in the definition, the conv4_3 layer shape should be `[None, 37, 37, some_channels]`  which is not the correct shape in ssd model(should be `[None, 38, 38, some_channels]`). adding an argumeng to change the padding from "valid" to "same" can solve this.

**Comment by reedwm** [comment]
I see, so you're feeding in larger images to vgg16, which causes each layer to have a larger height and width. And you want an option to change the padding of pool layers to "SAME" because the [SSD model](https://arxiv.org/abs/1512.02325) requires it when using vgg16 as a base.

@sguada what do you think of this proposal?

**Comment by sguada** [comment]
@zakizhou I'm not sure if you want to change the padding of the last max_pool layer of all of them, can you clarify what is the change you propose?

**Comment by zakizhou** [comment]
@sguada all pooling layers, not just the last one, for the first 3 pooling layers, "valid" padding works exactly same with "same" padding.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by sguada** [comment]
Having "VALID" for max_pooling with kernel size [2, 2] would change the size of the output, while "SAME" would not.

It seems that changing the polling of all max_poling layers would have different effects depending on the input size, so I guess it would be hard to control all possible cases.

You can build your own version of VGG with the padding that you need, isn't it?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?
